web_poet/utils.py

```diff
@@ -126,6 +133,63 @@ def new_method(self, *args, **kwargs):
     return new_method
 
 
+def cached_method(method):
+    """A decorator to cache method or coroutine method results,
+    so that if it's called multiple times for the same instance,
+    computation is only done once.
+
+    The cache is unbound, but it's tied to the instance lifetime.
+
+    .. note::
+
+        :func:`cached_method` is needed because :func:`functools.lru_cache`
+        doesn't work well on methods: self is used as a cache key,
+        so a reference to an instance is kept in the cache, and this
+        prevents deallocation of instances.
+    """
+    name_hash = hashlib.sha1(method.__name__.encode("utf8")).hexdigest()[:4]
+    cached_meth_name = f"_cached_{method.__name__}_{name_hash}"
```

Comment:
I read SHA-1 is quite fast, but why do we need a hash here in the first place?

Reply:
If we have:
```python
class Foo:
    def bar(self):
        print("hi")
```

Then `hashlib.sha1(Foo().bar.__name__.encode("utf-8")).hexdigest()[:4]` would always be `62cd`. In turn, `cached_meth_name` would have a value of `"_cached_bar_62cd'"`.

I'm also curious as to why the 4-character permutation at the end is needed, if they should be constant anyways.

Reply:
I added it as poor man's approach to avoid collisions with user-defined attributes, as an additional safeguard. You may have `self._cached_price` defined by user, but it's unlikely user defines `self._cached_price_62cd`.

Reply:
Speed shoudn't be an issue, as this code runs once, on import time.

Reply:
I think using `_cached_method_` as prefix and covering this in the documentation of the `cached_method` function should be enough safeguard. I mean, if anyone decorates `bar` with `@cached_method` and then defines `_cached_method_bar`, they are looking for trouble even without documentation warning them.